README.md
# flutter_map_text_marker

A [flutter_map](https://pub.dev/packages/flutter_map) plugin that adds the ability to create multiple text markers on a map.

## Getting started

To start using this package, make sure you have [flutter_map](https://pub.dev/packages/flutter_map) in your project's pubspec.yaml.

Then run the following command in your project's root directory:

```
flutter pub add flutter_map_text_marker
```

or add the following to your pubspec.yaml:

```
flutter_map_text_marker: ^0.0.3
```

Since the package implements `MapPlugin` from the `flutter_map` package, it has to be used within the FlutterMap widget as follows:

```dart
FlutterMap(
    mapController: controller,
    options: MapOptions(
        ...,
        plugins: [
            TextMarkerPlugin(),
        ],
    ),
    layers: [
        TileLayerOptions(...),
        TextMarkerPluginOptions(
            isActive: isActive,
            mapHeight: mapHeight,
            mapWidth: mapWidth,
            markers: markers,
            onAddMarker: onAddMarker,
            customAddDialog: (LatLng point) => getCustomAddDialog(point),
        ),
    ],
),
```

## Usage

Check out the example folder in this project to find a flutter application running with this plugin.

To add a text marker to the map, activate the "Add Marker" layer by clicking on the Floating Action Button, and long press on the position on the map where you want to add the marker.
You will be prompted to enter the text for the marker. Remember to exit the "Add Marker" layer by tapping on the Floating Action Button again.


CHANGELOG.md
## 0.0.3

* Added the ability to remove a `TextMarker`, with a confirmation SnackBar to the example project.
* Added support for using a custom Widget for the dialog to show when adding a new item.

## 0.0.2

* Cleaned up the code.
* Modified the plugin description inside the pubspec.yaml file.

## 0.0.1

* The initial release of this package adds the base functionality to add multiple text markers to a map.
* Future releases will improve overall code quality and add new features, such as customization of marker's text style, the ability to edit the marker's text and a customizable pop-up dialog to add/edit markers.


test/flutter_map_text_marker_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_map_text_marker/flutter_map_text_marker.dart';

void main() {
  test('tests', () {});
}


lib/flutter_map_text_marker.dart
import 'package:flutter/material.dart';
import 'package:latlong2/latlong.dart';
import 'package:flutter_map/plugin_api.dart';

class TextMarkerPluginOptions extends LayerOptions {
  List<TextMarker> markers;
  double mapWidth;
  double mapHeight;
  bool isActive;
  Function(TextMarker newMarker) onAddMarker;
  Function(LatLng point) onRemoveMarker;
  Widget Function(LatLng point)? customAddDialog;

  TextMarkerPluginOptions({
    this.markers = const [],
    required this.mapHeight,
    required this.mapWidth,
    required this.isActive,
    required this.onAddMarker,
    required this.onRemoveMarker,
    this.customAddDialog,
  });
}

class TextMarkerPlugin extends MapPlugin {
  @override
  Widget createLayer(
      LayerOptions options, MapState mapState, Stream<void> stream) {
    if (options is TextMarkerPluginOptions) {
      return StreamBuilder(
          stream: stream,
          builder: (BuildContext context, AsyncSnapshot<void> snapshot) {
            var textMarkers = <Widget>[];
            for (var marker in options.markers) {
              if (!_boundsContainsMarker(mapState, marker)) continue;

              textMarkers.add(TextMarkerWidget(
                mapState: mapState,
                marker: marker,
                stream: stream,
                options: options,
              ));
            }
            return TextMarkersOverlay(
              height: options.mapHeight,
              width: options.mapWidth,
              isActive: options.isActive,
              markers: textMarkers,
              mapState: mapState,
              onAddMarker: options.onAddMarker,
              onRemoveMarker: options.onRemoveMarker,
              customAddDialog: options.customAddDialog,
            );
          });
    }
    throw Exception('Unkown options type for type CustomPlugin'
        'plugin: $options');
  }

  @override
  bool supportsLayer(LayerOptions options) {
    return options is TextMarkerPluginOptions;
  }

  static bool _boundsContainsMarker(MapState map, TextMarker marker) {
    var pxPoint = map.project(marker.point);

    final width = marker.width - marker.anchor.left;
    final height = marker.height - marker.anchor.top;

    var sw = CustomPoint(pxPoint.x + width, pxPoint.y - height);
    var ne = CustomPoint(pxPoint.x - width, pxPoint.y + height);

    return map.pixelBounds.containsPartialBounds(Bounds(sw, ne));
  }
}

class TextMarkersOverlay extends StatefulWidget {
  const TextMarkersOverlay({
    Key? key,
    required this.height,
    required this.width,
    required this.isActive,
    required this.markers,
    required this.mapState,
    required this.onAddMarker,
    required this.onRemoveMarker,
    this.customAddDialog,
  }) : super(key: key);

  final double height;
  final double width;
  final bool isActive;
  final List<Widget> markers;
  final MapState mapState;
  final Function(TextMarker newMarker) onAddMarker;
  final Function(LatLng point) onRemoveMarker;
  final Widget Function(LatLng point)? customAddDialog;

  @override
  State<TextMarkersOverlay> createState() => _TextMarkersOverlayState();
}

class _TextMarkersOverlayState extends State<TextMarkersOverlay> {
  final TextEditingController _markerTextController = TextEditingController();

  @override
  void dispose() {
    _markerTextController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      decoration: BoxDecoration(
        border: Border.all(
          color: widget.isActive ? Colors.red : Colors.transparent,
          width: 1,
        ),
      ),
      child: GestureDetector(
        onLongPressEnd: (details) {
          if (!widget.isActive) return;
          // print("Long Press at ${details.localPosition}");
          LatLng markerPointPosition = _newMarkerCoords(details.localPosition);
          // Show dialog to ask user the text to insert into the text marker

          showDialog(
              context: context,
              builder: (context) {
                if (widget.customAddDialog != null) {
                  return widget.customAddDialog!(markerPointPosition);
                } else {
                  return Dialog(
                    backgroundColor: Colors.transparent,
                    child: Container(
                      height: 200,
                      width: 300,
                      decoration: BoxDecoration(
                        color: Colors.white,
                        borderRadius: BorderRadius.circular(10),
                      ),
                      child: Padding(
                        padding: const EdgeInsets.all(16.0),
                        child: Column(
                          children: [
                            const Text(
                              "Insert text for marker",
                              style: TextStyle(
                                  fontWeight: FontWeight.bold, fontSize: 18.0),
                              textAlign: TextAlign.center,
                            ),
                            Padding(
                              padding: const EdgeInsets.fromLTRB(
                                  20.0, 0.0, 20.0, 20.0),
                              child: TextField(
                                controller: _markerTextController,
                                decoration: const InputDecoration(
                                  border: UnderlineInputBorder(),
                                ),
                              ),
                            ),
                            SizedBox(
                              width: MediaQuery.of(context).size.width * 0.4,
                              child: ElevatedButton(
                                style: ButtonStyle(
                                  backgroundColor:
                                      MaterialStateProperty.all(Colors.blue),
                                ),
                                onPressed: () {
                                  // Here create and add the new marker to the list
                                  final String newWidgetText =
                                      _markerTextController.text;
                                  TextMarker newMarker = TextMarker(
                                    point: markerPointPosition,
                                    text: newWidgetText,
                                    onTap: (point) {
                                      // print("Marker at $point tapped");
                                    },
                                    onLongPress: (point) {
                                      // print("Marker at $point removed");
                                      widget.onRemoveMarker(point);
                                    },
                                    builder: (context) => Text(newWidgetText),
                                  );
                                  widget.onAddMarker(newMarker);
                                  _markerTextController.clear();
                                  Navigator.of(context).pop();
                                },
                                child: const Text("Add"),
                              ),
                            ),
                          ],
                        ),
                      ),
                    ),
                  );
                }
              });

          // print("Tapped coords: $markerPointPosition");
        },
        child: Container(
          width: widget.width,
          height: widget.height,
          decoration: BoxDecoration(
            color: widget.isActive
                ? Colors.black.withOpacity(0.5)
                : Colors.transparent,
          ),
          child: Stack(children: widget.markers),
        ),
      ),
    );
  }

  static CustomPoint _offsetToPoint(Offset offset) {
    return CustomPoint(offset.dx, offset.dy);
  }

  LatLng _newMarkerCoords(Offset offset) {
    MapState? mapState = widget.mapState;

    var renderObject = context.findRenderObject() as RenderBox;
    var width = renderObject.size.width;
    var height = renderObject.size.height;

    var localPoint = _offsetToPoint(offset);

    var localPointCenterDistance =
        CustomPoint((width / 2) - localPoint.x, (height / 2) - localPoint.y);

    var mapCenter = mapState.project(mapState.center);
    var point = mapCenter - localPointCenterDistance;
    return mapState.unproject(point);
  }
}

class TextMarkerWidget extends StatefulWidget {
  const TextMarkerWidget({
    Key? key,
    this.mapState,
    required this.marker,
    AnchorPos? anchorPos,
    this.stream,
    this.options,
  }) : super(key: key);

  final MapState? mapState;
  final TextMarker marker;
  final Stream<void>? stream;
  final LayerOptions? options;

  @override
  State<TextMarkerWidget> createState() => _TextMarkerWidgetState();
}

class _TextMarkerWidgetState extends State<TextMarkerWidget> {
  CustomPoint pixelPos = const CustomPoint(0.0, 0.0);
  late LatLng markerPos;

  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    TextMarker marker = widget.marker;
    updatePixelPos(widget.marker.point);

    return Positioned(
      left: pixelPos.x.toDouble(),
      top: pixelPos.y.toDouble(),
      child: GestureDetector(
        onTap: () {
          if (marker.onTap != null) {
            marker.onTap!(marker.point);
          }
        },
        onLongPress: () {
          if (marker.onLongPress != null) {
            marker.onLongPress!(marker.point);
          }
        },
        child: marker.builder!(context),
      ),
    );
  }

  void updatePixelPos(LatLng point) {
    TextMarker marker = widget.marker;
    MapState? mapState = widget.mapState;

    CustomPoint pos;
    if (mapState != null) {
      pos = mapState.project(point);
      pos =
          pos.multiplyBy(mapState.getZoomScale(mapState.zoom, mapState.zoom)) -
              mapState.getPixelOrigin();

      pixelPos = CustomPoint(
        (pos.x - (marker.width - widget.marker.anchor.left)).toDouble(),
        (pos.y - (marker.width - widget.marker.anchor.top)).toDouble(),
      );
    }
  }
}

class TextMarker {
  LatLng point;
  final WidgetBuilder? builder;
  final String text;
  final double width;
  final double height;
  final Offset offset;
  final Function(LatLng)? onTap;
  final Function(LatLng)? onLongPress;
  late Anchor anchor;

  TextMarker({
    required this.point,
    required this.text,
    this.builder,
    this.width = 30.0,
    this.height = 30.0,
    this.offset = const Offset(0.0, 0.0),
    this.onTap,
    this.onLongPress,
    AnchorPos? anchorPos,
  }) {
    anchor = Anchor.forPos(anchorPos, width, height);
  }
}


example/lib/map.dart
import 'package:flutter/material.dart';
import 'package:flutter_map/flutter_map.dart';
import 'package:latlong2/latlong.dart';
import 'package:flutter_map_text_marker/flutter_map_text_marker.dart';

class MapPage extends StatefulWidget {
  const MapPage({Key? key}) : super(key: key);

  @override
  State<MapPage> createState() => _MapPageState();
}

class _MapPageState extends State<MapPage> {
  MapController? _mapController;

  final List<TextMarker> _markers = [];

  void _addMarker(TextMarker newMarker) {
    // print("Adding marker: $newMarker");
    setState(() {
      _markers.add(newMarker);
    });
  }

  final TextEditingController _markerController = TextEditingController();

  Widget getCustomAddMarkerDialog(LatLng point) {
    return Dialog(
        child: Container(
      color: Colors.transparent,
      height: 200,
      width: 300,
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          TextField(
            controller: _markerController,
            decoration: const InputDecoration(
              labelText: "Marker text",
            ),
          ),
          ElevatedButton(
            child: const Text("Add"),
            onPressed: () {
              final String markerText = _markerController.text;
              _addMarker(
                TextMarker(
                    point: point,
                    text: markerText,
                    builder: (context) => Text(markerText),
                    onLongPress: (point) {
                      _removeMarker(point);
                    }),
              );
              Navigator.pop(context);
            },
          ),
        ],
      ),
    ));
  }

  void _removeMarker(LatLng point) {
    // print("Removing marker: $toDeleteMarker");
    TextMarker deletingMarker =
        _markers.firstWhere((marker) => marker.point == point);
    ScaffoldMessenger.of(_scaffoldKey.currentContext!).showSnackBar(
      SnackBar(
        content: Text(
          "Are you sure you want to remove this marker: ${deletingMarker.text}",
          style: const TextStyle(
            color: Colors.white,
          ),
        ),
        backgroundColor: Colors.red,
        action: SnackBarAction(
          label: "Yes, delete",
          textColor: Colors.white,
          onPressed: () {
            setState(() {
              _markers.removeWhere((marker) => marker.point == point);
            });
          },
        ),
      ),
    );
  }

  bool isTextMarkerActive = false;

  final GlobalKey _scaffoldKey = GlobalKey();

  @override
  Widget build(BuildContext context) {
    // print("Markers: $_markers");
    return Scaffold(
      key: _scaffoldKey,
      appBar: AppBar(
        title: const Text("Text Marker"),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          setState(() {
            isTextMarkerActive = !isTextMarkerActive;
          });
        },
        child: const Icon(Icons.edit),
      ),
      body: SizedBox(
        height: MediaQuery.of(context).size.height,
        width: MediaQuery.of(context).size.width,
        child: FlutterMap(
          mapController: _mapController,
          options: MapOptions(
            crs: const Epsg3857(),
            onMapCreated: (MapController controller) {
              _mapController = controller;
            },
            allowPanning: false,
            allowPanningOnScrollingParent: false,
            onPositionChanged: null,
            slideOnBoundaries: false,
            center: LatLng(41.889306777663066, 12.491833350660679),
            zoom: 18.0,
            plugins: [
              TextMarkerPlugin(),
            ],
          ),
          layers: [
            TileLayerOptions(
              urlTemplate: "https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png",
              subdomains: ['a', 'b', 'c'],
            ),
            TextMarkerPluginOptions(
              isActive: isTextMarkerActive,
              mapHeight: MediaQuery.of(context).size.height,
              mapWidth: MediaQuery.of(context).size.width,
              markers: _markers,
              onAddMarker: _addMarker,
              onRemoveMarker: _removeMarker,
              customAddDialog: (LatLng point) =>
                  getCustomAddMarkerDialog(point),
            ),
          ],
        ),
      ),
    );
  }
}
